### PR TITLE
Fix #21604, check Any fields at runtime

### LINKED ISSUE #21604: Spurious/incorrect method definition overwritten when struct defined with type alias of `Any` as field type inside module
When executing this code

```julia
module Z

const Foo = Any

struct X
    x::Foo
end

end
```

the spurious error here appears:

```julia
WARNING: Method definition (::Type{Z.X})(Any) in module Z at REPL[1]:6 overwritten at REPL[1]:6.
```

### PR DESCRIPTION
**Pull request by TotalVerb**
This moves the check for whether field types are `Any` or not to runtime instead of at lowering time.

I made the assumption that `X == Any` iff `X === Any` here; please tell me if that is invalid. (Maybe we need `Core.issubtype(Any, X)`?)

### CODE REVIEW

**Comment by martinholters** [comment]
Is it worth keeping the check in lowering to avoid the runtime check if it is known to be the case because there is a literal `Any` there?

**Comment by tkelman** [comment]
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

**Review by JeffBezanson — APPROVED** [review]

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/21607#issuecomment-297979561) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/52eecdb4fc7e749fa7f3a7f8132ef9ff19ec3a6f/348bfe9_vs_e6ab126/report.md). cc @jrevels

**Comment by TotalVerb** [comment]
Should we rerun the array and tuple benchmarks? I doubt the regressions are real.

**Comment by Sacha0** [comment]
> Should we rerun the array and tuple benchmarks? I doubt the regressions are real.

Some of the potential array regressions seem larger than typically occurs via noise, so to be safe @nanosoldier `runbenchmarks("array", vs = ":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/21607#issuecomment-298049889) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/1d4cf2c645a00f91e90233dd1a6857340a1fc189/1ee55e9_vs_de15be0/report.md). cc @jrevels

**Comment by TotalVerb** [comment]
The name looks similar but it's actually a different noisy `sumvector_view` benchmark, so I think this should be good to go.

**Comment by Sacha0** [comment]
Agreed, the combination of nanosoldier results look solid :).

**Comment by Sacha0** [comment]
@martinholters, your edit suggests you approve these changes? If so (with you and `JeffBezanson` both approving), perhaps merge at will? Best!

**Comment by martinholters** [comment]
Fine with me, but not that I have thoroughly reviewed this.

**Comment by JeffBezanson** [comment]
One thing we could do is use `(core Any)` as the default field type, and then we could skip the check for types `equal?` to that.

**Comment by TotalVerb** [comment]
@JeffBezanson I've implemented that suggestion. Could you check again?

**Comment by TotalVerb** [comment]
On a not-unrelated note, perhaps unnoticed until now because it is admittedly a very strange situation, there are some cases where we might prefer to use `(core Any)` instead of `'Any`:

```julia
julia> Any = Integer
WARNING: imported binding for Any overwritten in module Main
Integer

julia> type X end

julia> supertype(X)
Integer
```

**Comment by JeffBezanson** [comment]
Yes, might as well fix those too.

**Comment by TotalVerb** [comment]
I've replaced `'Any` with `'(core Any)` in many places and added some tests.

File: src/julia-syntax.scm [hunk 1]
```diff
@@ -164,7 +164,7 @@
                   `(block ,body))))
     `(lambda ,argl ()
              (scope-block
-              ,(if (eq? rett 'Any)
+              ,(if (eq? rett '(core Any))
```
Comment (JeffBezanson) [inline src/julia-syntax.scm]: This will need to use `equal?`.

**Review by JeffBezanson — CHANGES_REQUESTED** [review]

File: src/julia-syntax.scm (hunk 1 above)
Comment (TotalVerb) [inline src/julia-syntax.scm]: Good catch, thanks!

**Review by JeffBezanson — APPROVED** [review]

**Comment by Sacha0** [comment]
With the changes, what say you @nanosoldier `runbenchmarks(ALL, vs = ":master")` ?

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/21607#issuecomment-298176930) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/516cd0d72b84bbee80dcdc5474e28f1fc55028ab/7f340c3_vs_101dccc/report.md). cc @jrevels

**Comment by TotalVerb** [comment]
I think it's safe to say that these are noise.

**Pull request merged**